src/amber/router/pipe/session.cr

```diff
@@ -3,48 +3,77 @@ require "base64"
 require "yaml"
 require "openssl/hmac"
 
-module Amber::Pipe
-  # The session handler provides a cookie based session.  The handler will
-  # encode and decode the cookie and provide the hash in the context that can
-  # be used to maintain data across requests.
-  class Session < Base
-    property :key
-    property secret : String
-
-    def self.instance
-      @@instance ||= new
-    end
+module Amber
+  module Pipe
+    # The session handler provides a cookie based session.  The handler will
+    # encode and decode the cookie and provide the hash in the context that can
+    # be used to maintain data across requests.
+    class Session < Base
+      property :key
+      property secret : String
 
-    def initialize
-      @key = "#{Server.settings.project_name}.session"
-      @secret = Server.settings.secret
-    end
+      def self.instance
+        @@instance ||= new
+      end
 
-    def call(context : HTTP::Server::Context)
-      cookies = context.request.cookies
-      decode(context.session, cookies[@key].value) if cookies.has_key?(@key)
-      call_next(context)
-      value = encode(context.session)
-      cookies = context.response.cookies
-      cookies << HTTP::Cookie.new(@key, value)
-      cookies.add_response_headers(context.response.headers)
-      context
-    end
+      def initialize
+        @key = "#{Server.settings.project_name}.session"
+        @secret = Server.settings.secret
+      end
 
-    private def decode(session, data)
-      sha1, data = data.split("--", 2)
-      if sha1 == OpenSSL::HMAC.hexdigest(:sha1, @secret, data)
-        values = YAML.parse(Base64.decode_string(data))
-        values.each do |key, value|
-          session[key.to_s] = value.to_s
+      def call(context : HTTP::Server::Context)
+        cookies = context.request.cookies
+        decode(context.session, cookies[@key].value) if cookies.has_key?(@key)
+        call_next(context)
+        value = encode(context.session)
+        cookies = context.response.cookies
+        cookies << HTTP::Cookie.new(@key, value)
+        cookies.add_response_headers(context.response.headers)
+        context
+      end
+
+      private def decode(session, data)
+        sha1, data = data.split("--", 2)
+        if sha1 == OpenSSL::HMAC.hexdigest(:sha1, @secret, data)
+          values = YAML.parse(Base64.decode_string(data))
+          values.each do |key, value|
+            session[key.to_s] = value.to_s
+          end
         end
       end
+
+      private def encode(session)
+        data = Base64.encode(session.to_yaml)
+        sha1 = OpenSSL::HMAC.hexdigest(:sha1, @secret, data)
+        "#{sha1}--#{data}"
+      end
     end
+  end
+
+  module Router
+    module Session
```

Comment:
Maybe we can get rid of the `module Router` and just leave one module, for instance `module ContextSession`. I feel that we are nesting too many modules.